app/Topology.hs

```diff
@@ -21,23 +23,46 @@ data TopologyInfo = TopologyInfo {
   , topologyFile :: FilePath
   }
 
--- | IPv4 address with port number
+-- | IPv4 or IPv6 address with port number
 --
--- TODO: this type should be extended to take into account IPv6 addresses.
---
-data NodeAddress = NodeAddress HostName ServiceName
+data NodeAddress = NodeAddress {
+      naHostAddress :: !IP.IP
+    , naPort        :: !PortNumber
+    }
   deriving (Eq, Ord, Show)
 
-nodeAddressToSockAddr :: NodeAddress -> SockAddr
-nodeAddressToSockAddr (NodeAddress addr port) = SockAddrInet (read port) (IP.toHostAddress (read addr))
+nodeAddressToSockAddr
+    :: NodeAddress
+    -> SockAddr
+nodeAddressToSockAddr NodeAddress {naHostAddress, naPort} = case naHostAddress of
+    IP.IPv4 ipv4 -> SockAddrInet  naPort   (IP.toHostAddress  ipv4)
+    IP.IPv6 ipv6 -> SockAddrInet6 naPort 0 (IP.toHostAddress6 ipv6) 0
 
 instance Condense NodeAddress where
-    condense (NodeAddress addr port) = addr ++ ":" ++ show port
+    condense NodeAddress {naHostAddress, naPort}
+      = show naHostAddress ++ ":" ++ show naPort
 
 instance FromJSON NodeAddress where
-    parseJSON = withObject "NodeAddress" $ \v -> NodeAddress
-      <$> v .: "addr"
-      <*> v .: "port"
+    parseJSON = withObject "NodeAddress" $ \v -> do
+      NodeAddress
+      <$> (read <$> v .: "hostAddress")
```

Comment:
What was wrong with just `addr`? Look at the simple topology file for example. I'm not sure it's clearer. I'd probably stick to the old short name.

Reply:
I wanted something more explicit, and Karl suggested `HostAddress`, but ok let's switch to `addr`.